bolt-modules/boltlib/lib/puppet/functions/parallelize.rb

```diff
@@ -35,21 +35,15 @@ def parallelize(scope, data, &block)
     executor = Puppet.lookup(:bolt_executor)
     executor.report_function_call(self.class.name)
 
-    skein = data.each_with_index.map do |object, index|
-      executor.create_yarn(scope, block, object, index)
-    end
-
-    result = executor.round_robin(skein)
-
-    failed_indices = result.each_index.select do |i|
-      result[i].is_a?(Bolt::Error)
-    end
+    # Create a map of duplicates { "key" => 0 } to keep track of incrementing
+    # the number for the Fiber's name
+    duplicates = data.select { |e| data.count(e) > 1 }.to_a.uniq.product([0]).to_h
 
-    # TODO: Inner catch errors block?
-    if failed_indices.any?
-      raise Bolt::ParallelFailure.new(result, failed_indices)
+    futures = data.map do |object|
+      number = duplicates&.key?(object) ? duplicates[object] += 1 : ''
+      executor.create_fiber(scope, block, object, "Parallelize block for #{object}#{number}")
```

Comment:
I'm concerned this is a bad name but I don't know what better. For example if iterating over a series of hashes every log message would have the full hash..